[torch][amdsmi] Avoid ODR violation when loading amdsmi

danzimm opened the pull request:
Summary:
amdsmi bundles its own copy of `libamd_smi.so`. When you're interacting with `amdsmi` from *only* python that's fine, but when you try to interact with `libamd_smi.so` from native code too this poses a problem, because from native code you'll be linking against the copy of `libamd_smi.so` from the SDK.

This means you'll end up with 2 copies of `libamd_smi.so` in your process, and potentially (Murphey's law says you will, as does our CI) violate ODR.

In order to avoid this issue from the PT side of the world we can hook the `dlopen("path/to/bundled/libamd_smi.so")` and try to use the already loaded/SDK version of `libamd_smi.so` first, before proceeding to use the `path/to/bundled/libamd_smi.so`.

Test Plan: CI, inspect process using libamd_smi.so from native + python and observe only a single copy loaded

Differential Revision: D69064038

danzimm commented:
@pytorchbot label "topic: not user facing"

malfet commented on torch/cuda/__init__.py:
```diff
@@ -60,7 +60,37 @@
-            import amdsmi  # type: ignore[import]
+            import ctypes
+
+            _original_CDLL = ctypes.CDLL  # type: ignore[misc,assignment]
```
There is already a `torch._load_global_defs()` function that deals with preloading various dependencies for CUDA
https://github.com/pytorch/pytorch/blob/4c5a9a5f949ef3019fc3ef095034ccfc973ff13d/torch/__init__.py#L303

Why not `try: CDLL.dlopen("libami_smi.so");catch:` there and just leave a note here that this is done in order to avoid the situation? Because libraries are cached by their filename, so if dynamic linker finds there is already one instance of `libfoo.so` loaded it will return reference to that instance next time dlopen is called

danzimm replied:
Spoke offline, but inlining here too for history: `libamd_smi.so` is referred to by absolute path in `amdsmi`, so preloading won't work

malfet reviewed (approved):
Sure, it's a bit hacky (and imo would be nice to constrain to its own function), but looks good to me

malfet commented on torch/cuda/__init__.py:
```diff
@@ -60,7 +60,37 @@
+            # This creates an ODR violation if the copy of libamd_smi.so from lib
+            # has been loaded previously.
+            #
+            # In order to avoid this violation we hook CDLL and try using the
```
Would be good to add a permalink to https://github.com/ROCm/amdsmi/blob/9872083b491c9b136496fd493414aca5c6c144cb/py-interface/amdsmi_wrapper.py#L168 to explain why this logic is needed here

dmitrii-galantsev commented:
Will try to address in https://github.com/ROCm/amdsmi/issues/72 :)

danzimm committed:
[torch][amdsmi] Avoid ODR violation when loading amdsmi (#146324)

Summary:

amdsmi bundles its own copy of `libamd_smi.so`. When you're interacting with `amdsmi` from *only* python that's fine, but when you try to interact with `libamd_smi.so` from native code too this poses a problem, because from native code you'll be linking against the copy of `libamd_smi.so` from the SDK.

This means you'll end up with 2 copies of `libamd_smi.so` in your process, and potentially (Murphey's law says you will, as does our CI) violate ODR.

In order to avoid this issue from the PT side of the world we can hook the `dlopen("path/to/bundled/libamd_smi.so")` and try to use the already loaded/SDK version of `libamd_smi.so` first, before proceeding to use the `path/to/bundled/libamd_smi.so`.

Test Plan: CI, inspect process using libamd_smi.so from native + python and observe only a single copy loaded

Reviewed By: malfet

Differential Revision: D69064038

malfet commented:
@pytorrch merge

malfet commented:
@pytorchbot merge
